plugins/pipelines/app/models/concerns/samson_pipelines/stage_concern.rb

```diff
@@ -0,0 +1,37 @@
+module SamsonPipelines::StageConcern
+  # Return true if any stages in the pipeline are marked production
+  def production?
+    super || next_stages.any?(&:production?)
+  end
+
+  def next_stage
+    next_stage_ids.empty? ? super : Stage.find(next_stage_ids.first)
+  end
+
+  def next_stages
+    Stage.find(next_stage_ids)
+  end
+
+  protected
+
+  # Ensure we don't have a circular pipeline:
+  #
+  # potential race-condition if 2 stages are saved at same time:
+  #   stageA saved with pipelines to stageB and stageC
```

Comment:
you'd have to build up recursive calls to `project.stages.each(&:lock)` and run your validation in the final block